### PR TITLE
[Transform] Secondary credentials used with transforms should only require source and destination index privileges, not transform privileges

### PR DESCRIPTION
**Pull request by przemekwitek**
The internal `_validate` action was called by `PUT` action using secondary headers **only**. This was causing authorization issues if the API key (secondary header) was lacking `transform_admin` role.
This PR fixes this by providing **all** the security headers to `_validate` action using `ClientHelper.executeWithHeadersAsync` helper method rather than raw `client.execute`.

Relates https://github.com/elastic/elasticsearch/issues/93259

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportPutTransformAction.java
Comment (droberts195) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportPutTransformAction.java:124]: I find it surprising that this works. `config.setHeaders` was called inside a block that switched to secondary headers. So I would have thought that here we end up called `ValidateTransformAction` using the secondary headers, which should fail if those secondary headers only allow index access and not transform administration.

I can see you have a test that seems to test that this functionality works, but please can you do some extra due diligence on whether this is really working as expected, perhaps by running the test locally with extra debug, or by manually testing it with `gradle run`.

If this extra testing shows that the existing test is only succeeding by some fluke then I think the way to fix it would be to put this call back to how it was before, and instead switch to secondary headers (using `useSecondaryAuthIfAvailable`) within the subsections of `TransportValidateTransformAction` that access the source or destination indices.

If the code really does work correctly as you've got it now then I need to adjust my understanding of what `executeWithHeadersAsync` does.

Comment (przemekwitek) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportPutTransformAction.java]: > So I would have thought that here we end up calling ValidateTransformAction using the secondary headers, which should fail if those secondary headers only allow index access and not transform administration.

Yes, that was indeed the case and the scenario from Quynh I reproduced.

> I can see you have a test that seems to test that this functionality works, but please can you do some extra due diligence on whether this is really working as expected, perhaps by running the test locally with extra debug, or by manually testing it with gradle run.

So, I did a little back and forth on this PR and the version you reviewed probably had the test (my new test written exactly for this case) failing.

The reason for this was I noticed Ben's work (https://github.com/elastic/elasticsearch/pull/86757) solving the issue https://github.com/elastic/elasticsearch/issues/86731 (Prefer secondary authentication headers when creating/updating transforms) and I wanted to keep his tests passing as well.

I ended up with the following (this PR is now updated):
- pass **all** the security headers to `_validate` call
- only store **secondary** headers in the transform config.

This way `_validate` call can be executed even if the secondary headers does not have `transform_admin`.
And at the same time we only store secondary header in the config in order to use it for data access later on (e.g: creating destination index).

**Comment by przemekwitek** [comment]
Some tests are now failing on CI.
So it seems this PR needs more work. I'll update it once it's ready for review again.

**Comment by przemekwitek** [comment]
> Some tests are now failing on CI. So it seems this PR needs more work. I'll update it once it's ready for review again.

Ok, the tests are fixed. It was a simple NPE causing these failures.
PR is ready for review again.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStartTransformAction.java
Comment (droberts195) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStartTransformAction.java:243]: This is a change in behaviour. Currently you don't need to supply secondary headers when starting a transform, but this change would mean that you would have to.

I think the way around it is to change the approach to calling `ValidateTransformAction` completely.

We should call `ValidateTransformAction` using the internal `_xpack` user, i.e. `executeAsyncWithOrigin`. (`ValidateTransformAction` doesn't even have a REST endpoint - from the end user's perspective it doesn't exist. It is entirely an internal implementation detail.) Then, internally, `TransportValidateTransformAction` should execute client calls using `config.getHeaders()`.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransformUpdater.java
Comment (droberts195) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransformUpdater.java:215]: Use `executeAsyncWithOrigin`. Then none of the other changes in this file are needed.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportPutTransformAction.java
Comment (droberts195) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportPutTransformAction.java:128]: Use `executeAsyncWithOrigin`.

**Review by droberts195 — COMMENTED** [review]
Unfortunately I still think there is a problem.

I think the solution is to call `_validate` using the internal `_xpack` user, as it's entirely an internal endpoint.

Then, in `TransportValidateTransformAction` the functions that use a `Client` should switch to the appropriate headers instead of relying on the caller of `_validate` to have done this. As far as I can see that's: https://github.com/elastic/elasticsearch/blob/a8a684ebabe678aa2a661ba5047bb2a5e45e17a6/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportValidateTransformAction.java#L130
and: https://github.com/elastic/elasticsearch/blob/a8a684ebabe678aa2a661ba5047bb2a5e45e17a6/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportValidateTransformAction.java#L139

**Comment by przemekwitek** [comment]
> Unfortunately I still think there is a problem.
> 
> I think the solution is to call `_validate` using the internal `_xpack` user, as it's entirely an internal endpoint.

Done.

> Then, in `TransportValidateTransformAction` the functions that use a `Client` should switch to the appropriate headers instead of relying on the caller of `_validate` to have done this. As far as I can see that's:
> 
> https://github.com/elastic/elasticsearch/blob/a8a684ebabe678aa2a661ba5047bb2a5e45e17a6/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportValidateTransformAction.java#L130
> 
> 
> and:
> https://github.com/elastic/elasticsearch/blob/a8a684ebabe678aa2a661ba5047bb2a5e45e17a6/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportValidateTransformAction.java#L139

Done.
I call these actions using `executeWithHeadersAsync` and provide headers stored in the config (`config.getHeaders()`).

**Review by droberts195 — APPROVED** [review]
LGTM if you could just rename a method as per https://github.com/elastic/elasticsearch/pull/94420#discussion_r1132532710 before merging.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/SecondaryAuthorizationUtils.java
Comment (przemekwitek) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/SecondaryAuthorizationUtils.java]: Done.

**Pull request merged**
